AP_HAL: Add serial option to not override streamrates

stephendade opened the pull request:
This PR adds an option to SERIALn_OPTIONS (4096) for ArduPilot to _not_ process any ``REQUEST_DATA_STREAM`` messages.

This is useful where the vehicle is operating over a low-datarate telemetry link and the ``SRn_*`` parameters are deliberately set to reduce the vehicle's telemetry traffic.

In this case, a GCS configured (by default) with a streamrate of 4 will then override the ``SRn_*`` parameters via the ``REQUEST_DATA_STREAM`` message and potentially flood the telemetry link.

Of course, all GCS's can be configured to _not_ set any streamrates, but this PR makes it explicit and lowers this risk of any human error in (mis)configuring their GCS.

This PR also includes a small fix to AP_HAL_ChibiOS that fixes the UART get_options() to use the correct variable type.

Tested in SITL and a Pixhawk1-1M

khancyr reviewed (commented):
wonderfull !
LGTM

peterbarker commented on libraries/AP_HAL/UARTDriver.h:
```diff
@@ -61,8 +63,8 @@ class AP_HAL::UARTDriver : public AP_HAL::BetterStream {
     // control optional features
-    virtual bool set_options(uint16_t options) { return options==0; }
-    virtual uint8_t get_options(void) const { return 0; }
+    virtual bool set_options(uint16_t options) { _last_options = options; return options; }
```
You've inverted the response of this method.

Somebody better have a good reason for this not being `void`....

stephendade replied:
Fixed

olliw42 commented:
I would think that this would be a great feature ... if it would do what is being said it is doing, namely to ignore stream rate changes from GCSes ... but this is not what it is doing, it is ignoring all altogether. This can be quite an issue if you do have e.g. a companion which sets the stream rates.

What I think one wants is indeed what was claimed, namely that GCSes are blocked out from changing the streamrates,  but not other components. Ergo, I think there should be a check of who is requesting the change. (could be based on MAV_TYPE, or different sys id, or sys id equal to sysid of my gcs)

Note that REQUEST_DATA_STREAM is deprecated and replaced by MAV_CMD_SET_MESSAGE_INTERVAL, see https://mavlink.io/en/messages/common.html#REQUEST_DATA_STREAM (the description seems to have a typo)
Note that there are therefore several paths were similar changes might be needed.

Note also that the stream rate request can be targeted. One can argue that the blocking-out should only apply to broadcast requestes, but not to targeted requests.

Hwurzburg commented:
I tried a while back to get an option bit to prevent GCS override of SRx params and the concept was rejected...the same logic should apply here.....or the option reconsidered (it still causes issues with users of LR telem)

stephendade commented:
I've updated to fix up the parameter generation.

Note I've changed the name of the PR to better describe what I'm trying to do: Ignore _any_ streamrate changes that come in over a telemetry port.

My reasoning for this goes back to what I'm trying to protect against: accidental flooding of a low-bandwidth telemetry link by a misconfigured system on the link - usually a GCS (but not always).

tridge commented on libraries/AP_HAL/UARTDriver.h:
```diff
@@ -15,6 +15,8 @@ class AP_HAL::UARTDriver : public AP_HAL::BetterStream {
     UARTDriver(const UARTDriver &other) = delete;
     UARTDriver &operator=(const UARTDriver&) = delete;
 
+    uint16_t _last_options;
```
should be private

rmackay9 replied:
maybe stick a comment beside it as well..

stephendade replied:
Fixed
